File: py/pr_landing_planner/plan.py
"""PURE fusion: combine the conflict graph + CI verdicts + freshness + priority into a landing PLAN.

This is the missing fusion the predecessor tools never did: neither the conflict-graph tools nor the
CI-health tool ever computed "these PRs are conflict-free, all green, all fresh — land them now,
these reds are benign, refire them, this one is a real regression, hold it." :func:`compute_plan`
does exactly that, deterministically, from already-collected data.

Per-PR action assignment (the fusion table):

* held (base-conflict)          -> ``rebase-then-land``
* held (draft / depends-on-held)-> ``wait``
* gate-policy change            -> ``escalate-gate-policy``
* exact-head local evidence     -> ``land-now`` (or rebase first), without a merge-gate wait
* CI runner-outage              -> ``escalate-runner-outage``
* CI evaluate-once race         -> ``wait`` (benign; treat as pending)
* CI stale required check       -> ``refire-stale-gate``
* CI flaky red                  -> ``refire-ci``
* CI real red                   -> ``hold-fix``
* CI pending / no checks        -> ``wait``
* CI green but behind base      -> ``rebase-then-land``
* CI green, fresh, gate ok      -> ``land-now``

Ordering among actionable PRs follows the parallel-safe layering, which itself ranks by
priority -> diff size -> age -> PR number.
"""

from __future__ import annotations

from collections.abc import Sequence

from pr_landing_planner.graph import build_stacks, held_reasons, partition_parallel_safe
from pr_landing_planner.model import (
    CiState,
    CollectedGraph,
    ConflictEdge,
    Diagnostics,
    HeldPr,
    OrderingEdge,
    Plan,
    PlanResult,
    PrAction,
    PrActionDecision,
    PrNode,
    PolicyClass,
    RedClass,
    ValidationEvidence,
)

DEFAULT_FRESHNESS_MAX_BEHIND = 0


def _held_action(reasons: Sequence[str]) -> tuple[PrAction, str]:
    if any(r in ("local-base-conflict", "github-base-conflicting") for r in reasons):
        return PrAction.REBASE_THEN_LAND, f"held: {', '.join(reasons)} — rebase onto base to resolve"
    if any(r.startswith("depends-on-held") for r in reasons):
        return PrAction.WAIT, f"held: {', '.join(reasons)}"
    if "draft" in reasons:
        return PrAction.WAIT, "held: draft (mark ready to land)"
    return PrAction.WAIT, f"held: {', '.join(reasons)}"


def _ci_action(node: PrNode, freshness_max_behind: int) -> tuple[PrAction, str]:
    ci = node.ci
    red = ci.red_class
    if node.policy_class is PolicyClass.GATE_POLICY:
        return (
            PrAction.ESCALATE_GATE_POLICY,
            "gate-policy change requires coordinator decision; validation evidence is not approval",
        )
    if node.validation_evidence in (
        ValidationEvidence.LOCALLY_VALIDATED,
        ValidationEvidence.CLEAN_VALIDATE_RECORD,
    ):
        if node.commits_behind > freshness_max_behind:
            return (
                PrAction.REBASE_THEN_LAND,
                f"{node.validation_evidence.value} at exact head; "
                f"rebase {node.commits_behind} commit(s) then land without waiting for merge-gate",
            )
        return (
            PrAction.LAND_NOW,
            f"{node.validation_evidence.value} at exact head; no merge-gate wait",
        )
    if red is RedClass.RUNNER_OUTAGE:
        return PrAction.ESCALATE_RUNNER_OUTAGE, ci.detail
    if red is RedClass.EVALUATE_ONCE_RACE:
        return PrAction.WAIT, ci.detail
    if red is RedClass.STALE_REQUIRED_CHECK:
        return PrAction.REFIRE_STALE_GATE, ci.detail
    if red is RedClass.FLAKY:
        return PrAction.REFIRE_CI, ci.detail
    if red is RedClass.REAL:
        return PrAction.HOLD_FIX, ci.detail
    if ci.raw_state is CiState.PENDING:
        return PrAction.WAIT, "CI pending"
    if ci.raw_state is CiState.NONE:
        return PrAction.WAIT, "no CI checks configured"
    # GREEN.
    if node.commits_behind > freshness_max_behind:
        return (
            PrAction.REBASE_THEN_LAND,
            f"green but {node.commits_behind} commit(s) behind base",
        )
    return PrAction.LAND_NOW, "authoritative CI green, fresh, gate ok"


def _greedy_conflict_free(
    numbers: Sequence[int],
    conflicts: dict[int, set[int]],
    rank: dict[int, tuple[int, int, str, int]],
) -> tuple[int, ...]:
    chosen: list[int] = []
    for number in sorted(numbers, key=lambda n: rank[n]):
        if all(peer not in conflicts.get(number, set()) for peer in chosen):
            chosen.append(number)
    return tuple(chosen)


def compute_plan(
    nodes: Sequence[PrNode],
    conflict_edges: Sequence[ConflictEdge],
    ordering_edges: Sequence[OrderingEdge],
    held: Sequence[HeldPr],
    *,
    freshness_max_behind: int = DEFAULT_FRESHNESS_MAX_BEHIND,
    outage_min_prs: int = 2,
    batch: bool = False,
) -> tuple[Plan, Diagnostics]:
    """Fuse everything into a :class:`Plan` + :class:`Diagnostics` (pure; deterministic)."""
    held_by_number = {h.pr: h for h in held}
    held_set = frozenset(held_by_number)

    groups = partition_parallel_safe(nodes, conflict_edges, ordering_edges, exclude=held_set)
    group_of: dict[int, int] = {
        number: idx for idx, group in enumerate(groups) for number in group
    }

    decisions: list[PrActionDecision] = []
    for node in sorted(nodes, key=lambda n: n.number):
        if node.number in held_by_number:
            action, why = _held_action(held_by_number[node.number].reasons)
        else:
            action, why = _ci_action(node, freshness_max_behind)
        decisions.append(
            PrActionDecision(
                pr=node.number, action=action, why=why, group=group_of.get(node.number)
            )
        )

    land_now = tuple(d.pr for d in decisions if d.action is PrAction.LAND_NOW)
    # Recommended act sequence: flatten the parallel-safe groups (already priority-ranked per layer).
    order = tuple(number for group in groups for number in group)

    # Optional bors-style batch: one conflict-free set of already-green land-now PRs behind one gate.
    batch_prs: tuple[int, ...] = ()
    if batch and land_now:
        conflicts: dict[int, set[int]] = {}
        for edge in conflict_edges:
            conflicts.setdefault(edge.a, set()).add(edge.b)
            conflicts.setdefault(edge.b, set()).add(edge.a)
        rank = {
            n.number: (n.priority, n.size, n.created_at, n.number)
            for n in nodes
        }
        batch_prs = _greedy_conflict_free(land_now, conflicts, rank)

    plan = Plan(
        parallel_safe_groups=groups,
        land_now=land_now,
        order=order,
        per_pr_actions=tuple(decisions),
        batch=batch_prs,
    )

    stale_gates = tuple(
        n.number for n in nodes if n.ci.red_class is RedClass.STALE_REQUIRED_CHECK
    )
    flaky_reds = tuple(n.number for n in nodes if n.ci.red_class is RedClass.FLAKY)
    real_reds = tuple(n.number for n in nodes if n.ci.red_class is RedClass.REAL)
    evaluate_once = tuple(
        n.number for n in nodes if n.ci.red_class is RedClass.EVALUATE_ONCE_RACE
    )
    outage_prs = tuple(n.number for n in nodes if n.ci.red_class is RedClass.RUNNER_OUTAGE)
    diagnostics = Diagnostics(
        stale_gates=stale_gates,
        flaky_reds=flaky_reds,
        real_reds=real_reds,
        evaluate_once_race=evaluate_once,
        outage_prs=outage_prs,
        outage_suspected=len(outage_prs) >= outage_min_prs,
    )
    return plan, diagnostics


def assemble_result(
    graph: CollectedGraph,
    *,
    freshness_max_behind: int = DEFAULT_FRESHNESS_MAX_BEHIND,
    outage_min_prs: int = 2,
    batch: bool = False,
) -> PlanResult:
    """Pure end-to-end assembly: a collected graph -> stacks + held PRs + the fused plan + diagnostics."""
    stacks = build_stacks(graph.ordering_edges)
    held = held_reasons(graph.nodes, graph.ordering_edges)
    plan, diagnostics = compute_plan(
        graph.nodes,
        graph.conflict_edges,
        graph.ordering_edges,
        held,
        freshness_max_behind=freshness_max_behind,
        outage_min_prs=outage_min_prs,
        batch=batch,
    )
    return PlanResult(
        graph=graph, stacks=stacks, held=held, plan=plan, diagnostics=diagnostics
    )


__all__ = ["compute_plan", "assemble_result", "DEFAULT_FRESHNESS_MAX_BEHIND"]

File: rs/pr-landing-planner/src/plan.rs
//! Pure fusion of graph, freshness, evidence, and CI into recommended actions.

use std::collections::{BTreeMap, BTreeSet};

use crate::graph::{build_stacks, held_reasons, partition_parallel_safe};
use crate::model::{
    CiState, CollectedGraph, ConflictEdge, Diagnostics, HeldPr, OrderingEdge, Plan, PlanResult,
    PolicyClass, PrAction, PrActionDecision, PrNode, RedClass, ValidationAuthority,
    ValidationEvidence,
};

/// Default freshness policy. `None` leaves base age advisory unless the caller opts in.
pub const DEFAULT_FRESHNESS_MAX_BEHIND: Option<i64> = None;

fn held_action(reasons: &[String]) -> (PrAction, String) {
    if reasons.iter().any(|reason| reason == "ordering-cycle") {
        return (
            PrAction::Wait,
            "held: ordering-cycle (resolve dependency cycle and rerun)".to_owned(),
        );
    }
    if reasons.iter().any(|reason| {
        matches!(reason.as_str(), "review-required" | "changes-requested")
            || reason.starts_with("review-decision-unknown:")
            || reason.starts_with("review-pass-")
    }) {
        return (PrAction::Wait, format!("held: {}", reasons.join(", ")));
    }
    if reasons
        .iter()
        .any(|reason| reason.starts_with("depends-on-held"))
    {
        return (PrAction::Wait, format!("held: {}", reasons.join(", ")));
    }
    if reasons.iter().any(|reason| reason == "draft") {
        return (
            PrAction::Wait,
            "held: draft (mark ready to land)".to_owned(),
        );
    }
    if reasons
        .iter()
        .any(|reason| reason == "local-base-conflict" || reason == "github-base-conflicting")
    {
        return (
            PrAction::RebaseThenLand,
            format!(
                "held: {} — rebase onto base, then revalidate before landing",
                reasons.join(", ")
            ),
        );
    }
    (PrAction::Wait, format!("held: {}", reasons.join(", ")))
}

fn ci_action(node: &PrNode, freshness_max_behind: Option<i64>) -> (PrAction, String) {
    if node.policy_class == PolicyClass::GatePolicy {
        return (
            PrAction::EscalateGatePolicy,
            "gate-policy change requires coordinator decision; validation evidence is not approval"
                .to_owned(),
        );
    }
    if node.validation_evidence == ValidationEvidence::CleanValidateRecord {
        if node.validation_authority == ValidationAuthority::None {
            return (
                PrAction::Wait,
                "clean-validate-record has no consuming-workspace hard/soft-green authority"
                    .to_owned(),
            );
        }
        return (
            PrAction::LandNow,
            format!(
                "{} at exact head with {} authority; no merge-gate wait",
                node.validation_evidence.as_str(),
                node.validation_authority.as_str(),
            ),
        );
    }
    if let Some(red) = node.ci.red_class {
        let action = match red {
            RedClass::RunnerOutage => PrAction::EscalateRunnerOutage,
            RedClass::EvaluateOnceRace => PrAction::Wait,
            RedClass::StaleRequiredCheck => PrAction::RefireStaleGate,
            RedClass::Flaky => PrAction::RefireCi,
            RedClass::Real => PrAction::HoldFix,
        };
        return (action, node.ci.detail.clone());
    }
    if !node.ci.gate_present {
        return (
            PrAction::RefireCi,
            "required gate has NO_RESULT (absent); re-dispatch".to_owned(),
        );
    }
    if !node.ci.gate_ok {
        return (
            PrAction::RefireCi,
            "required gate has NO_RESULT; re-dispatch".to_owned(),
        );
    }
    if node.ci.raw_state == CiState::NoResult {
        return (
            PrAction::Wait,
            "required gate passed; another CI check has NO_RESULT".to_owned(),
        );
    }
    if freshness_max_behind.is_some_and(|limit| node.commits_behind > limit) {
        return (
            PrAction::RebaseThenLand,
            format!("green but {} commit(s) behind base", node.commits_behind),
        );
    }
    (
        PrAction::LandNow,
        "authoritative CI green, gate ok".to_owned(),
    )
}

fn greedy_conflict_free(
    numbers: &[i64],
    conflicts: &BTreeMap<i64, BTreeSet<i64>>,
    by_number: &BTreeMap<i64, &PrNode>,
) -> Vec<i64> {
    let mut sorted = numbers.to_vec();
    sorted.sort_by_key(|number| {
        let node = by_number[number];
        (
            node.priority,
            node.size(),
            node.created_at.as_str(),
            node.number,
        )
    });
    let mut chosen = Vec::new();
    for number in sorted {
        if chosen
            .iter()
            .all(|peer| !conflicts.get(&number).is_some_and(|set| set.contains(peer)))
        {
            chosen.push(number);
        }
    }
    chosen
}

/// Fuse graph, CI, holds, evidence, freshness, and dependencies into recommendations.
pub fn compute_plan(
    nodes: &[PrNode],
    conflict_edges: &[ConflictEdge],
    ordering_edges: &[OrderingEdge],
    held: &[HeldPr],
    freshness_max_behind: Option<i64>,
    outage_min_prs: usize,
    batch: bool,
) -> (Plan, Diagnostics) {
    let held_by_number: BTreeMap<_, _> = held.iter().map(|held| (held.pr, held)).collect();
    let held_set = held_by_number.keys().copied().collect();
    let groups = partition_parallel_safe(nodes, conflict_edges, ordering_edges, &held_set);
    let group_of: BTreeMap<_, _> = groups
        .iter()
        .enumerate()
        .flat_map(|(group, numbers)| numbers.iter().map(move |number| (*number, group)))
        .collect();

    let mut sorted_nodes: Vec<_> = nodes.iter().collect();
    sorted_nodes.sort_by_key(|node| node.number);
    let mut decisions = sorted_nodes
        .into_iter()
        .map(|node| {
            let (action, why) = if node.policy_class == PolicyClass::GatePolicy {
                ci_action(node, freshness_max_behind)
            } else if let Some(held) = held_by_number.get(&node.number) {
                held_action(&held.reasons)
            } else {
                ci_action(node, freshness_max_behind)
            };
            PrActionDecision {
                pr: node.number,
                action,
                why,
                group: group_of.get(&node.number).copied(),
            }
        })
        .collect::<Vec<_>>();
    let mut by_action: BTreeMap<_, _> = decisions
        .iter()
        .map(|decision| (decision.pr, decision.action))
        .collect();
    let mut blockers = ordering_edges.to_vec();
    blockers.sort_by_key(|edge| (edge.after, edge.before));
    loop {
        let mut changed = false;
        for decision in &mut decisions {
            let blocker = blockers.iter().find_map(|edge| {
                (edge.after == decision.pr
                    && by_action.get(&edge.before) != Some(&PrAction::LandNow))
                .then_some(edge.before)
            });
            if let Some(blocker) = blocker {
                if matches!(
                    decision.action,
                    PrAction::LandNow | PrAction::RebaseThenLand
                ) {
                    let action = by_action.get(&blocker).copied().unwrap_or(PrAction::Wait);
                    decision.action = PrAction::Wait;
                    decision.why = format!(
                        "dependency #{blocker} requires {}; wait and rerun",
                        action.as_str()
                    );
                    if by_action.insert(decision.pr, PrAction::Wait) != Some(PrAction::Wait) {
                        changed = true;
                    }
                }
            }
        }
        if !changed {
            break;
        }
    }
    let land_now = decisions
        .iter()
        .filter(|decision| decision.action == PrAction::LandNow)
        .map(|decision| decision.pr)
        .collect::<Vec<_>>();
    let order = groups.iter().flatten().copied().collect();
    let mut batch_prs = Vec::new();
    if batch && !land_now.is_empty() {
        let mut conflicts: BTreeMap<i64, BTreeSet<i64>> = BTreeMap::new();
        for edge in conflict_edges {
            conflicts.entry(edge.a).or_default().insert(edge.b);
            conflicts.entry(edge.b).or_default().insert(edge.a);
        }
        let by_number = nodes.iter().map(|node| (node.number, node)).collect();
        let ordered_children: BTreeSet<_> = ordering_edges.iter().map(|edge| edge.after).collect();
        let candidates = land_now
            .iter()
            .filter(|number| !ordered_children.contains(number))
            .copied()
            .collect::<Vec<_>>();
        batch_prs = greedy_conflict_free(&candidates, &conflicts, &by_number);
    }
    let red = |class| {
        nodes
            .iter()
            .filter(|node| node.ci.red_class == Some(class))
            .map(|node| node.number)
            .collect::<Vec<_>>()
    };
    let outage_prs = red(RedClass::RunnerOutage);
    let diagnostics = Diagnostics {
        stale_gates: red(RedClass::StaleRequiredCheck),
        flaky_reds: red(RedClass::Flaky),
        real_reds: red(RedClass::Real),
        evaluate_once_race: red(RedClass::EvaluateOnceRace),
        outage_suspected: outage_prs.len() >= outage_min_prs,
        outage_prs,
    };
    (
        Plan {
            parallel_safe_groups: groups,
            land_now,
            order,
            per_pr_actions: decisions,
            batch: batch_prs,
        },
        diagnostics,
    )
}

/// Derive stacks and holds, then compute a complete result from a collected graph.
pub fn assemble_result(
    graph: CollectedGraph,
    freshness_max_behind: Option<i64>,
    outage_min_prs: usize,
    batch: bool,
) -> PlanResult {
    let stacks = build_stacks(&graph.ordering_edges);
    let held = held_reasons(&graph.nodes, &graph.ordering_edges);
    let (plan, diagnostics) = compute_plan(
        &graph.nodes,
        &graph.conflict_edges,
        &graph.ordering_edges,
        &held,
        freshness_max_behind,
        outage_min_prs,
        batch,
    );
    PlanResult {
        graph,
        stacks,
        held,
        plan,
        diagnostics,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{CiVerdict, PolicyClass};

    fn green(number: i64) -> PrNode {
        PrNode {
            number,
            ci: CiVerdict {
                raw_state: CiState::Passed,
                gate_present: true,
                gate_ok: true,
                detail: "passed".into(),
                ..CiVerdict::default()
            },
            ..PrNode::default()
        }
    }

    #[test]
    fn fusion_table_respects_evidence_policy_and_freshness() {
        let mut evidence = green(1);
        evidence.validation_evidence = ValidationEvidence::CleanValidateRecord;
        evidence.validation_authority = ValidationAuthority::HardGreen;
        evidence.ci = CiVerdict::default();
        let mut policy = green(2);
        policy.policy_class = PolicyClass::GatePolicy;
        let mut behind = green(3);
        behind.commits_behind = 1;
        let (plan, _) = compute_plan(&[evidence, policy, behind], &[], &[], &[], None, 2, false);
        assert_eq!(plan.per_pr_actions[0].action, PrAction::LandNow);
        assert_eq!(plan.per_pr_actions[1].action, PrAction::EscalateGatePolicy);
        assert_eq!(plan.per_pr_actions[2].action, PrAction::LandNow);

        let mut strict_behind = green(4);
        strict_behind.commits_behind = 1;
        let (strict, _) = compute_plan(&[strict_behind], &[], &[], &[], Some(0), 2, false);
        assert_eq!(strict.per_pr_actions[0].action, PrAction::RebaseThenLand);

        let mut unproven = green(5);
        unproven.validation_evidence = ValidationEvidence::CleanValidateRecord;
        let (unproven_plan, _) = compute_plan(&[unproven], &[], &[], &[], None, 2, false);
        assert_eq!(unproven_plan.per_pr_actions[0].action, PrAction::Wait);
        assert!(unproven_plan.per_pr_actions[0]
            .why
            .contains("no consuming-workspace hard/soft-green authority"));
    }

    #[test]
    fn passing_non_gate_ci_never_substitutes_for_required_gate() {
        let missing = PrNode {
            number: 1,
            ci: CiVerdict {
                raw_state: CiState::Passed,
                gate_present: false,
                gate_ok: false,
                ..CiVerdict::default()
            },
            ..PrNode::default()
        };
        let non_passing = PrNode {
            number: 2,
            ci: CiVerdict {
                raw_state: CiState::Passed,
                gate_present: true,
                gate_ok: false,
                ..CiVerdict::default()
            },
            ..PrNode::default()
        };
        let (plan, _) = compute_plan(&[missing, non_passing], &[], &[], &[], None, 2, false);
        assert_eq!(plan.per_pr_actions[0].action, PrAction::RefireCi);
        assert_eq!(plan.per_pr_actions[1].action, PrAction::RefireCi);
        assert!(plan.land_now.is_empty());
    }

    #[test]
    fn dependencies_block_landable_children_and_batches_exclude_children() {
        let parent = PrNode {
            number: 1,
            ci: CiVerdict {
                raw_state: CiState::Failed,
                red_class: Some(RedClass::Real),
                gate_present: true,
                detail: "real red".into(),
                ..CiVerdict::default()
            },
            ..PrNode::default()
        };
        let child = green(2);
        let edge = OrderingEdge {
            before: 1,
            after: 2,
            reason: "base-ref".into(),
        };
        let (blocked, _) = compute_plan(
            &[parent, child],
            &[],
            std::slice::from_ref(&edge),
            &[],
            None,
            2,
            true,
        );
        let child = blocked
            .per_pr_actions
            .iter()
            .find(|decision| decision.pr == 2)
            .unwrap();
        assert_eq!(child.action, PrAction::Wait);
        assert!(child.why.contains("dependency #1 requires hold-fix"));
        assert!(!blocked.land_now.contains(&2));
        assert!(!blocked.batch.contains(&2));

        let (ready, _) = compute_plan(
            &[green(3), green(4)],
            &[],
            &[OrderingEdge {
                before: 3,
                after: 4,
                reason: "base-ref".into(),
            }],
            &[],
            None,
            2,
            true,
        );
        assert_eq!(ready.land_now, vec![3, 4]);
        assert_eq!(ready.batch, vec![3]);
    }

    #[test]
    fn gate_policy_escalates_and_main_advance_preserves_authorized_exact_head_evidence() {
        let mut policy = green(1);
        policy.policy_class = PolicyClass::GatePolicy;
        let held = HeldPr {
            pr: 1,
            reasons: vec!["draft".into()],
        };
        let (policy_plan, _) = compute_plan(&[policy], &[], &[], &[held], None, 2, false);
        assert_eq!(
            policy_plan.per_pr_actions[0].action,
            PrAction::EscalateGatePolicy
        );

        let mut evidence = green(2);
        evidence.validation_evidence = ValidationEvidence::CleanValidateRecord;
        evidence.validation_authority = ValidationAuthority::SoftGreen;
        evidence.commits_behind = 3;
        let (evidence_plan, _) = compute_plan(&[evidence], &[], &[], &[], None, 2, false);
        let decision = &evidence_plan.per_pr_actions[0];
        assert_eq!(decision.action, PrAction::LandNow);
        assert!(decision.why.contains("soft-green authority"));
        assert!(!decision.why.contains("rebase"));
    }

    #[test]
    fn all_red_actions_no_result_variants_batch_and_diagnostics() {
        let red_node = |number, class| PrNode {
            number,
            ci: CiVerdict {
                raw_state: CiState::Failed,
                red_class: Some(class),
                gate_present: true,
                detail: class.as_str().into(),
                ..CiVerdict::default()
            },
            ..PrNode::default()
        };
        let mut nodes = vec![
            red_node(1, RedClass::StaleRequiredCheck),
            red_node(2, RedClass::Flaky),
            red_node(3, RedClass::Real),
            red_node(4, RedClass::EvaluateOnceRace),
            red_node(5, RedClass::RunnerOutage),
            red_node(6, RedClass::RunnerOutage),
            PrNode {
                number: 7,
                ci: CiVerdict {
                    raw_state: CiState::NoResult,
                    ..CiVerdict::default()
                },
                ..PrNode::default()
            },
            PrNode {
                number: 8,
                ci: CiVerdict {
                    raw_state: CiState::NoResult,
                    gate_present: true,
                    gate_ok: true,
                    ..CiVerdict::default()
                },
                ..PrNode::default()
            },
            green(9),
            green(10),
            green(11),
        ];
        nodes[8].priority = 1;
        nodes[9].priority = 0;
        nodes[10].priority = 0;
        nodes[10].additions = 1;
        let conflicts = vec![ConflictEdge {
            a: 9,
            b: 10,
            paths: vec!["x".into()],
        }];
        let (plan, diagnostics) = compute_plan(&nodes, &conflicts, &[], &[], None, 2, true);
        let actions: BTreeMap<_, _> = plan
            .per_pr_actions
            .iter()
            .map(|decision| (decision.pr, decision.action))
            .collect();
        assert_eq!(actions[&1], PrAction::RefireStaleGate);
        assert_eq!(actions[&2], PrAction::RefireCi);
        assert_eq!(actions[&3], PrAction::HoldFix);
        assert_eq!(actions[&4], PrAction::Wait);
        assert_eq!(actions[&5], PrAction::EscalateRunnerOutage);
        assert_eq!(actions[&7], PrAction::RefireCi);
        assert_eq!(actions[&8], PrAction::Wait);
        assert!(diagnostics.outage_suspected);
        assert_eq!(diagnostics.outage_prs, vec![5, 6]);
        assert!(!(plan.batch.contains(&9) && plan.batch.contains(&10)));
        assert!(plan.batch.contains(&11));
    }
}
